File: P4Ejercicio9/Java/controlador/Controlador.java
import javax.swing.*;
import javax.swing.table.DefaultTableModel;
import java.awt.event.ActionEvent;
import java.awt.event.ActionListener;
import java.io.BufferedReader;
import java.io.FileReader;
import java.util.ArrayList;
import java.util.Arrays;

public class Controlador {
    private CSV modelo;
    private Vista vista;
    private String rutaArchivo;

    public Controlador(CSV modelo, Vista vista) {
        this.modelo = modelo;
        this.vista = vista;

        // Action listeners for buttons
        vista.getBotonSeleccionarArchivo().addActionListener(new ActionListener() {
            @Override
            public void actionPerformed(ActionEvent e) {
                seleccionarArchivo();
            }
        });

        vista.getBotonLeerArchivo().addActionListener(new ActionListener() {
            @Override
            public void actionPerformed(ActionEvent e) {
                cargarArchivo();
            }
        });

        this.rutaArchivo = null;
    }

    public Vista getVista()
    {
        return vista;
    }

    private void seleccionarArchivo() {
        JFileChooser fileChooser = new JFileChooser();
        fileChooser.setDialogTitle("Seleccionar archivo CSV");
        int result = fileChooser.showOpenDialog(vista);
        if (result == JFileChooser.APPROVE_OPTION) {
            rutaArchivo = fileChooser.getSelectedFile().getAbsolutePath();
            vista.getMensajeError().setText("Archivo seleccionado: " + rutaArchivo);
        }
    }

    private void cargarArchivo() {
        if (rutaArchivo == null) {
            vista.getMensajeError().setText("Primero debe seleccionar un archivo.");
            return;
        }

        try {
            leerArchivo(rutaArchivo);
            mostrarDatosEnTabla();
            vista.getMensajeError().setText("");
        } catch (Exception e) {
            vista.getMensajeError().setText(e.getMessage());
        }
    }

    private void leerArchivo(String rutaArchivo) throws Exception {
        try (BufferedReader br = new BufferedReader(new FileReader(rutaArchivo))) {
            String encabezado = br.readLine();
            if (!encabezado.equals(String.join(",", DataLabels.FIELDS))) {
                throw new Exception("El archivo no tiene el formato de encabezado correcto.");
            }

            String linea;

            while ((linea = br.readLine()) != null) {
                String[] datos = linea.split(",");

                if (datos.length != DataLabels.FIELDS.size()) {
                    throw new Exception("Línea no cumple con el formato esperado: " + linea);
                }

                for (int i = 0; i < DataLabels.FIELDS.size(); i++) {
                    modelo.getData().add(new ArrayList<String>(Arrays.asList(DataLabels.FIELDS.get(i), datos[i])));
                }
            }
        }
    }

    private void mostrarDatosEnTabla() {
        DefaultTableModel tableModel = (DefaultTableModel) vista.getTableModel();
        tableModel.setRowCount(0);  // Limpiar la tabla

        for (ArrayList<String> fila : modelo.getData()) {
            Object[] rowData = new Object[DataLabels.FIELDS.size()];
            for (int i = 0; i < DataLabels.FIELDS.size(); i++) {
                rowData[i] = fila.get(i);
            }
            tableModel.addRow(rowData);
        }
    }
}

File: P4Ejercicio9/Python/controlador/Controlador.py
import csv
from PyQt6.QtWidgets import QFileDialog, QMessageBox, QTableWidgetItem
from modelo.DataLabels import DataLabels

class Controlador:
    def __init__(self, modelo, vista):
        self.modelo = modelo
        self.vista = vista
        
        self.vista.boton_seleccionar_archivo.clicked.connect(self.seleccionar_archivo)
        self.vista.boton_leer_archivo.clicked.connect(self.cargar_archivo)

        self.ruta_archivo = None

    def seleccionar_archivo(self):
        self.ruta_archivo, _ = QFileDialog.getOpenFileName(self.vista, "Seleccionar archivo CSV", "", "CSV Files (*.csv)")
        if self.ruta_archivo:
            self.vista.mensaje_error.setText("Archivo seleccionado: " + self.ruta_archivo)

    def cargar_archivo(self):
        if not self.ruta_archivo:
            self.vista.mensaje_error.setText("Primero debe seleccionar un archivo.")
            return
        
        try:
            self.leer_archivo(self.ruta_archivo)
            self.mostrar_datos_en_tabla()
            self.vista.mensaje_error.setText("")
        except Exception as e:
            self.vista.mensaje_error.setText(str(e))

    def leer_archivo(self, ruta_archivo):
        try:
            with open(ruta_archivo, newline='', encoding='utf-8') as archivo_csv:
                lector = csv.reader(archivo_csv)
                encabezado = next(lector)
                if encabezado != DataLabels.FIELDS:
                    raise ValueError("El archivo no tiene el formato de encabezado correcto.")

                for linea in lector:
                    if len(linea) != len(DataLabels.FIELDS):
                        raise ValueError(f"Línea no cumple con el formato esperado: {linea}")
                    
                    fila_dict = {encabezado[i]: linea[i] for i in range(len(encabezado))}
                    self.modelo.data.append(fila_dict)

        except Exception as e:
            raise e

    def mostrar_datos_en_tabla(self):
        self.vista.tabla.setRowCount(len(self.modelo.data))
        
        for fila, dato in enumerate(self.modelo.data):
            for i in range(len(DataLabels.FIELDS)):
                self.vista.tabla.setItem(fila, i, QTableWidgetItem(dato[DataLabels.FIELDS[i]]))
